cpp/src/arrow/util/crc32.cc

```diff
@@ -0,0 +1,961 @@
+ * including commercial applications, and to alter it and redistribute it
+ * freely, subject to the following restrictions:
+ *
+ * 1. The origin of this software must not be misrepresented; you must not
+ *    claim that you wrote the original software.
+ * 2. If you use this software in a product, an acknowledgment in the product
+ *    documentation would be appreciated but is not required.
+ * 3. Altered source versions must be plainly marked as such, and must not be
+ *    misrepresented as being the original software.
+ *
+ * Taken from http://create.stephan-brumme.com/crc32/ and adapted for Cyrus
+ * by Robert Norris <robn@fastmail.com>
+ */
+
+/*
+ * Copyright (c) 1994-2016 Carnegie Mellon University.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * 3. The name "Carnegie Mellon University" must not be used to
+ *    endorse or promote products derived from this software without
+ *    prior written permission. For permission or any legal
+ *    details, please contact
+ *      Carnegie Mellon University
+ *      Center for Technology Transfer and Enterprise Creation
+ *      4615 Forbes Avenue
+ *      Suite 302
+ *      Pittsburgh, PA  15213
+ *      (412) 268-7393, fax: (412) 268-7395
+ *      innovation@andrew.cmu.edu
+ *
+ * 4. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by Computing Services
+ *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
+ *
+ * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
+ * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
+ * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+ * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
+ * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "arrow/util/crc32.h"
+
+#include <cstdint>
+
+#include "arrow/util/endian.h"
+
+namespace arrow {
+namespace internal {
+
+#define ALIGNOF_UINT32_T alignof(uint32_t)
+
+static const uint32_t crc32_lookup[16][256] = {
+    /* same algorithm as crc32_bitwise
+
+       for (int i = 0; i <= 0xFF; i++)
+       {
+         uint32_t crc = i;
+         for (int j = 0; j < 8; j++)
+           crc = (crc >> 1) ^ ((crc & 1) * Polynomial);
```

Comment:
Looks below table is generated with Polynomial = 0xEDB88320, which is CRC32.
CPU normally implements CRC32**C** instructions with Polynomial = 0x82F63B78.
Can we first try the [tests ](https://github.com/apache/arrow/pull/14351#discussion_r1082367597)mentions by @pitrou to see which one is preferred?

Reply:
No, standards says we need CRC32 https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L642-L659

Reply:
AFAIK, Intel CPU only supports CRC32C, Arm supports both.

Reply:
Yes, but parquet standard already uses crc32. If crc32c is used, there are so many well done libraries that we can use.